### PR TITLE
Add some working examples

### PR DESCRIPTION
**Pull request by craigtaub**
### Description of the Change

Added working examples for a couple of common configurations. More to come if these are useful.

This includes
- programmatic usage
- react, webpack
- basic typescript

### Alternate Designs

More complex examples? - i decided it was best to start simple and focus on the setup, easier to expand after that.

Thinking about how these could be added to the documentation website. Perhaps just link in docs for now.

### Benefits

- help contributors quickly setup scenario to debug (quick way to confirm a bug), 
- help users have up-to-date examples of multiple given setups
- can be used to demonstrate common and complex functionality

### Applicable issues

https://github.com/mochajs/mocha/issues/2359

### CODE REVIEW

**Comment by craigtaub** [comment]
@juergba 

> Our examples could focus on set-up's, Wiki on Mocha's features. 

With the Wiki covering so many features that might be a good divide. I like the idea of having working examples for features which are covered enough by documentation (not wiki) but still give users issues.

> Anyway, I would prefer those examples on the seperate repo, especially if it is going to grow in size. 

At this point I dont quite agree, heres why
- danger with a separate repo is that its quite detached and reliant on people finding the link (im just speculating)
- dont think we should worry about it growing in size, its not downloaded by clients and monorepo's are typically huge. `examples` folder is pretty self-contained.
- code dies quickly in separate repo. Inside the main one it will be very visible and (more likely) used and kept up-to-date. 
- the other feedback I received was put it inside `examples`.

Thoughts? 

> "npm install" will not distribute these examples, right?

Nope, by default wont install subdirs.

> Are the "package-lock.json" necessary to include?

I thought as a best practice it was best to keep, plus consistent + deterministic installations would make it easier for people to get started (w/o worry about updates/fixes themselves). Not completely convinced. What do you think?

> Topics I have missed in the past:

Good idea I can add those to the list 👍

**Comment by plroebuck** [comment]
> Anyway, I would prefer those examples on the seperate repo, especially if it is going to grow in size. With a link in our documentation.

Agree. See no point in having larger distribution package due to example code.
Think these should be in another repository.

File: .eslintignore [hunk 1]
```diff
@@ -6,3 +6,4 @@ out/
 !lib/mocha.js
 test/integration/fixtures
 !.*.js
+example/applications
```
Comment (plroebuck) [inline .eslintignore]: Why are examples immune to `prettier`?

File: example/applications/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Example Applications
+
+A set of working examples to help demontrate common configurations using Mocha.
+
+## Requirements
+
+Each example application will have its own package dependencies but also a [version of Node.js](https://docs.npmjs.com/files/package.json#engines) it was designed and built from. Lower versions have not been tested so might not work, please beware. Feel free to submit any fixes for lower versions.
```
Comment (plroebuck) [inline example/applications/README.md]: "beware" -> "be aware"?

File: example/applications/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+## Reporting a bug?
+
+Feel free to send a branch to supply an MCVE's from one of the applications.
```
Comment (plroebuck) [inline example/applications/README.md]: MCVE's -> MCVE

File: example/applications/programmatic-usage/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (plroebuck) [inline example/applications/programmatic-usage/.gitignore]: Why is this file even here? Irrelevant

File: example/applications/programmatic-usage/index.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  reporter: 'list'
+});
+
+var testDir = 'tests/';
+
+// Add each .js file to the mocha instance
+fs.readdirSync(testDir)
+  .filter(function(file) {
+    // Only keep the .js files
+    return file.substr(-3) === '.js';
```
Comment (plroebuck) [inline example/applications/programmatic-usage/index.js]: prefer higher-level abstraction...
```js
return path.extname(file) === '.js';
```

File: example/applications/programmatic-usage/index.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+var fs = require('fs');
+
+var path = require('path');
+
+// Instantiate a Mocha
+var mocha = new Mocha({
+  reporter: 'list'
+});
+
+var testDir = 'tests/';
```
Comment (plroebuck) [inline example/applications/programmatic-usage/index.js]: Was point to use non-default directory? If so, comment that.

File: example/applications/programmatic-usage/tests/simple.spec.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+const { equal } = require("assert");
+const index = require('../src/index');
+
+describe('programmatic usage suite', function() {
+
+  it('should pass', function() {
+    // pass
+  });
```
Comment (plroebuck) [inline example/applications/programmatic-usage/tests/simple.spec.js]: Pointless. Make it do something or remove.

File: example/applications/programmatic-usage/tests/simple.spec.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+const { equal } = require("assert");
+const index = require('../src/index');
```
Comment (plroebuck) [inline example/applications/programmatic-usage/tests/simple.spec.js]:
```js
const index = require('../src');
```

File: example/applications/react-webpack/package.json [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    "@babel/register": "^7.0.0",
+    "babel-loader": "^8.0.5",
+    "chai": "^4.2.0",
+    "enzyme": "^3.9.0",
+    "enzyme-adapter-react-16": "^1.10.0",
+    "mocha": "^6.0.2",
+    "react": "^16.8.3",
+    "react-dom": "^16.8.3",
+    "webpack": "^4.29.5"
+  },
```
Comment (plroebuck) [inline example/applications/react-webpack/package.json]: Some of these "dependencies" should be "devDependencies" instead...

File: example/applications/react-webpack/package.json [hunk 10]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "example-applications-react-webpack",
+  "version": "1.0.0",
+  "description": "React Webpack example",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha",
+    "build": "webpack",
+    "start": "npm run build && open index.html"
```
Comment (plroebuck) [inline example/applications/react-webpack/package.json]: "open" is a binary on macOS, but definitely not available on all platforms...

File: example/applications/typescript/src/hello/world.spec.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+import world from './world';
+
+describe('#World', () => {
+  it('should be able to execute a test', function() {
+    equal(true, true);
+  });
+  it('should return expected string', function() {
+    equal(world('incoming'), "incoming-static")
+  });
+});
```
Comment (plroebuck) [inline example/applications/typescript/src/hello/world.spec.ts]: Why isn't this file in "test" directory?

File: example/applications/typescript/package.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    "test": "mocha",
+    "compile": "tsc"
+  },
+  "engines": {
+    "node": ">=10.0.0"
+  },
+  "license": "ISC",
+  "dependencies": {
+    "@types/mocha": "^5.2.6",
+    "mocha": "^6.0.0",
```
Comment (plroebuck) [inline example/applications/typescript/package.json]: devDependency

File: example/applications/typescript/tsconfig.json [hunk 13]
```diff
@@ -0,0 +1,8 @@
+{
+  "compilerOptions": {
+    "outDir": "./dist",
```
Comment (plroebuck) [inline example/applications/typescript/tsconfig.json]: Isn't canonical use to transpile to "lib" directory?

File: example/applications/typescript/README.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
+# Typescript application
+
+Full documentation about it [here](https://mochajs.org/#-require-module-r-module)
+
+Tests here are kept next to their code (not in a separate dir).
```
Comment (plroebuck) [inline example/applications/typescript/README.md]: Explain why done this way.

File: example/applications/typescript/README.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+
+## npm scripts
```
Comment (plroebuck) [inline example/applications/typescript/README.md]: Can't say I see why this section is included in "README"s as it's self-apparent from "package.json". Suggest instead to convert to set of shell commands used to show how to use example.

**Comment by craigtaub** [comment]
Have made comments inside PR https://github.com/mochajs/mocha-working-examples/pull/1. Closing this as its now located there.

**Pull request closed without merging**